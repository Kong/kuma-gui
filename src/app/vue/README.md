---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

