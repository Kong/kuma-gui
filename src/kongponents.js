import Vue from 'vue'
import '@kongponents/styles/styles.css'
import KAlert from '@kongponents/kalert'
import KBadge from '@kongponents/kbadge'
import KButton from '@kongponents/kbutton'
import KCard from '@kongponents/kcard'
import KEmptyState from '@kongponents/kemptystate'
import KIcon from '@kongponents/kicon'
import KTable from '@kongponents/ktable'
import KToaster from '@kongponents/ktoaster'
import KoolTip from '@kongponents/kooltip'
import KClipboardProvider from '@kongponents/kclipboardprovider'
import Krumbs from '@kongponents/krumbs'
import KPop from '@kongponents/kpop'
import KToggle from '@kongponents/ktoggle'
import KTabs from '@kongponents/ktabs'
import KModal from '@kongponents/kmodal'
import KInput from '@kongponents/kinput'
import KInputSwitch from '@kongponents/kinputswitch'

Vue.component('KAlert', KAlert)
Vue.component('KBadge', KBadge)
Vue.component('KButton', KButton)
Vue.component('KCard', KCard)
Vue.component('KEmptyState', KEmptyState)
Vue.component('KIcon', KIcon)
Vue.component('KTable', KTable)
Vue.component('KToaster', KToaster)
Vue.component('KoolTip', KoolTip)
Vue.component('KClipboardProvider', KClipboardProvider)
Vue.component('Krumbs', Krumbs)
Vue.component('KPop', KPop)
Vue.component('KTabs', KTabs)
Vue.component('KToggle', KToggle)
Vue.component('KInput', KInput)
Vue.component('KInputSwitch', KInputSwitch)
Vue.component('KModal', KModal)
