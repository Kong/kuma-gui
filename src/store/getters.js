export const workspace = (state) => state.workspace

export const getSelectedMesh = (state) => {
  return state.selectedMesh
}
