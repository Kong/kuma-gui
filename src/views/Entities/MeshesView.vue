<template>
  <div class="all-meshes">
    <FrameSkeleton>
      <DataOverview
        :page-size="pageSize"
        :has-error="hasError"
        :is-loading="isLoading"
        :empty-state="empty_state"
        :table-data="tableData"
        :table-data-is-empty="tableDataIsEmpty"
        :next="next"
        @table-action="tableAction"
        @load-data="loadData($event)"
      >
        <template #additionalControls>
          <KButton
            class="add-mesh-button"
            appearance="primary"
            size="small"
            :to="{ path: '/wizard/mesh' }"
            @click="onCreateClick"
          >
            <span class="custom-control-icon">
              +
            </span>
            Create Mesh
          </KButton>
        </template>
      </DataOverview>
      <TabsWidget
        v-if="isEmpty === false"
        :has-error="hasError"
        :is-loading="isLoading"
        :tabs="tabs"
        initial-tab-override="overview"
      >
        <template #tabHeader>
          <div v-if="entity.basicData">
            <h3> Mesh: {{ entity.basicData.name }}</h3>
          </div>
        </template>
        <template #overview>
          <LabelList
            :has-error="entityHasError"
            :is-loading="entityIsLoading"
            :is-empty="entityIsEmpty"
          >
            <div>
              <ul>
                <li
                  v-for="(value, key) in entity.basicData"
                  :key="key"
                >
                  <h4 v-if="key === 'creationTime'">
                    Created
                  </h4>
                  <h4 v-else-if="key === 'modificationTime'">
                    Last Modified
                  </h4>
                  <h4 v-else>
                    {{ key }}
                  </h4>
                  <p v-if="key === 'creationTime' || key === 'modificationTime'">
                    {{ readableDate(value) }} <em>({{ rawDate(value) }})</em>
                  </p>
                  <p v-else>
                    {{ value }}
                  </p>
                </li>
              </ul>
            </div>
            <div v-if="entity.extendedData && entity.extendedData.length">
              <ul>
                <li
                  v-for="(item, key) in entity.extendedData"
                  :key="key"
                >
                  <h4>{{ item.label }}</h4>
                  <p
                    v-if="item.value"
                    class="label-cols"
                  >
                    <span>
                      {{ item.value.type }}
                    </span>
                    <span>
                      {{ item.value.name }}
                    </span>
                  </p>
                  <KBadge
                    v-else
                    appearance="danger"
                  >
                    Disabled
                  </KBadge>
                </li>
                <li>
                  <h4>Locality Aware Loadbalancing</h4>
                  <p v-if="entity.localityEnabled">
                    <KBadge appearance="success">
                      Enabled
                    </KBadge>
                  </p>
                  <KBadge
                    v-else
                    appearance="danger"
                  >
                    Disabled
                  </KBadge>
                </li>
              </ul>
            </div>
          </LabelList>
        </template>
        <template #yaml>
          <YamlView
            :has-error="entityHasError"
            :is-loading="entityIsLoading"
            :is-empty="entityIsEmpty"
            :content="rawEntity"
          />
        </template>
        <template #resources>
          <LabelList
            :has-error="entityHasError"
            :is-loading="entityIsLoading"
            :is-empty="entityIsEmpty"
          >
            <div
              v-for="i in countCols"
              :key="i"
            >
              <ul>
                <li
                  v-for="(item, key) in counts.slice((i - 1) * itemsPerCol, i * itemsPerCol)"
                  :key="key"
                >
                  <h4>{{ item.title }}</h4>
                  <p>{{ formatValue(item.value) }}</p>
                </li>
              </ul>
            </div>
          </LabelList>
        </template>
      </TabsWidget>
    </FrameSkeleton>
  </div>
</template>

<script>
import { mapGetters, mapState } from 'vuex'
import { datadogLogs } from '@datadog/browser-logs'

import Kuma from '@/services/kuma'
import { getTableData } from '@/utils/tableDataUtils'
import { getEmptyInsight } from '@/store/reducers/mesh-insights'
import { datadogLogEvents } from '@/datadogEvents'
import { getSome, humanReadableDate, rawReadableDate, stripTimes } from '@/helpers'
import FrameSkeleton from '@/components/Skeletons/FrameSkeleton.vue'
import DataOverview from '@/components/Skeletons/DataOverview.vue'
import TabsWidget from '@/components/Utils/TabsWidget.vue'
import YamlView from '@/components/Skeletons/YamlView.vue'
import LabelList from '@/components/Utils/LabelList.vue'
import { PAGE_SIZE_DEFAULT } from '@/consts'

export default {
  name: 'MeshesView',

  components: {
    FrameSkeleton,
    DataOverview,
    TabsWidget,
    YamlView,
    LabelList,
  },

  data() {
    return {
      isLoading: true,
      isEmpty: false,
      hasError: false,
      entityIsLoading: true,
      entityIsEmpty: false,
      entityHasError: false,
      tableDataIsEmpty: false,
      empty_state: {
        title: 'No Data',
        message: 'There are no Meshes present.',
      },
      tableData: {
        headers: [
          { key: 'actions', hideLabel: true },
          { label: 'Name', key: 'name' },
          { label: 'Type', key: 'type' },
        ],
        data: [],
      },
      tabs: [
        {
          hash: '#overview',
          title: 'Overview',
        },
        {
          hash: '#resources',
          title: 'Resources',
        },
        {
          hash: '#yaml',
          title: 'YAML',
        },
      ],
      entity: {},
      rawEntity: {},
      pageSize: PAGE_SIZE_DEFAULT,
      next: null,
      itemsPerCol: 3,
      meshInsight: getEmptyInsight(),
    }
  },
  computed: {
    ...mapState({
      policies: (state) => state.policies,
    }),

    ...mapGetters({
      featureFlags: 'config/featureFlags',
    }),
    counts() {
      const policies = this.policies.map((policy) => ({
        title: policy.pluralDisplayName,
        value: this.meshInsight.policies[policy.name]?.total || 0,
      }))

      return [
        {
          title: 'Data plane proxies',
          value: this.meshInsight.dataplanes.total,
        },
        ...policies,
      ]
    },
    countCols() {
      return Math.ceil(this.counts.length / this.itemsPerCol)
    },
  },
  watch: {
    $route() {
      // Ensures basic state is reset when switching meshes using the mesh selector.
      this.isLoading = true
      this.isEmpty = false
      this.hasError = false
      this.entityIsLoading = true
      this.entityIsEmpty = false
      this.entityHasError = false
      this.tableDataIsEmpty = false

      this.loadData()
    },
  },
  beforeMount() {
    this.init()
  },
  methods: {
    init() {
      this.loadData()
    },

    onCreateClick() {
      datadogLogs.logger.info(datadogLogEvents.CREATE_MESH_CLICKED)
    },
    tableAction(ev) {
      const data = ev

      // load the data into the tabs
      this.getEntity(data)
    },
    async loadData(offset = '0') {
      this.isLoading = true
      this.isEmpty = false

      const mesh = this.$route.params.mesh
      let query

      if (mesh !== 'all') {
        query = this.$route.params.mesh
      }

      try {
        const { data, next } = await getTableData({
          getSingleEntity: Kuma.getMesh.bind(Kuma),
          getAllEntities: Kuma.getAllMeshes.bind(Kuma),
          size: this.pageSize,
          offset,
          query,
        })

        // set pagination
        this.next = next

        // set table data
        if (data.length) {
          this.tableData.data = [...data]
          this.tableDataIsEmpty = false
          this.isEmpty = false

          this.getEntity({ name: data[0].name })
        } else {
          this.tableData.data = []
          this.tableDataIsEmpty = true
          this.isEmpty = true
          this.entityIsEmpty = true
        }
      } catch (error) {
        this.hasError = true
        this.isEmpty = true

        console.error(error)
      } finally {
        this.isLoading = false
      }
    },
    getEntity(entity) {
      this.entityIsLoading = true
      this.entityIsEmpty = false
      this.entityHasError = false

      if (entity && entity !== null) {
        return Kuma.getMesh({ name: entity.name })
          .then((response) => {
            if (response) {
              Kuma.getMeshInsights({ name: entity.name }).then((meshInsightResponse) => {
                this.meshInsight = meshInsightResponse
              })

              // get the counts for this mesh

              // const col1 = getSome(response, ['type', 'name', 'creationTime', 'modificationTime'])
              const col1 = getSome(response, ['type', 'name'])

              const formatted = () => {
                const data = Object.entries(getSome(response, ['mtls', 'logging', 'metrics', 'tracing']))
                const newData = []

                data.forEach((i) => {
                  const label = i[0]
                  const subData = i[1] || null

                  /**
                   * If `enabledBackend` is present, this will match it against
                   * the name found in `backends`, get the `type` from there, and
                   * output the defaultBackend name alongside its matched `type`.
                   *
                   * For Tracing, `defaultBackend` is used instead of `enabledBackend`.
                   */

                  if (subData && subData.enabledBackend) {
                    const enabled = subData.enabledBackend
                    const matched = subData.backends.find((obj) => obj.name === enabled)

                    if (matched) {
                      newData.push({
                        label,
                        value: {
                          type: matched.type,
                          name: matched.name,
                        },
                      })
                    }
                  } else if (subData && subData.defaultBackend) {
                    const enabled = subData.defaultBackend
                    const matched = subData.backends.find((obj) => obj.name === enabled)

                    if (matched) {
                      newData.push({
                        label,
                        value: {
                          type: matched.type,
                          name: matched.name,
                        },
                      })
                    }
                  } else if (subData && subData.backends) {
                    const backends = subData.backends[0]

                    if (backends) {
                      newData.push({
                        label,
                        value: {
                          type: backends.type,
                          name: backends.name,
                        },
                      })
                    }
                  } else {
                    newData.push({
                      label,
                      value: null,
                    })
                  }
                })

                return newData
              }

              const isRoutingEnabled = () => {
                const { routing } = response

                return routing && routing.localityAwareLoadBalancing
              }

              this.entity = {
                basicData: col1,
                extendedData: formatted(),
                localityEnabled: isRoutingEnabled(),
              }

              // this.rawEntity = response
              this.rawEntity = stripTimes(response)
            } else {
              this.entity = {}
              this.entityIsEmpty = true
            }
          })
          .catch((error) => {
            this.entityHasError = true
            console.error(error)
          })
          .finally(() => {
            setTimeout(() => {
              this.entityIsLoading = false
            }, import.meta.env.VITE_DATA_TIMEOUT)
          })
      } else {
        setTimeout(() => {
          this.entityIsEmpty = true
          this.entityIsLoading = false
        }, import.meta.env.VITE_DATA_TIMEOUT)
      }
    },

    formatValue(value) {
      return value ? value.toLocaleString('en').toString() : 0
    },

    readableDate(value) {
      return humanReadableDate(value)
    },

    rawDate(value) {
      return rawReadableDate(value)
    },
  },
}
</script>

<style lang="scss" scoped>
.add-mesh-button {
  background-color: var(--logo-green) !important;
}
</style>
