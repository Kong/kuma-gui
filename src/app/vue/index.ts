import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
