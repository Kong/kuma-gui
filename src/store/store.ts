import { InjectionKey } from 'vue'
import { createStore, Store } from 'vuex'

import { storeConfig, State } from './index'

export const storeKey: InjectionKey<Store<State>> = Symbol('store')

export const store = createStore<State>(storeConfig)
