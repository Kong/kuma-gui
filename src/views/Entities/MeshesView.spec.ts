import { RouterLinkStub } from '@vue/test-utils'
import { render, screen } from '@testing-library/vue'
import userEvent from '@testing-library/user-event'
import { KAlert, KBadge, KButton, KCard, KClipboardProvider, KEmptyState, KIcon, KPop, KTable, KTabs } from '@kong/kongponents'

import MeshesView from './MeshesView.vue'
import { store, storeKey } from '@/store/store'

describe('MeshesView.vue', () => {
  it('renders snapshot with Resource tab', async () => {
    await store.dispatch('fetchPolicies')

    const { container } = render(MeshesView, {
      global: {
        plugins: [[store, storeKey]],
        components: { KAlert, KBadge, KButton, KCard, KClipboardProvider, KEmptyState, KIcon, KPop, KTable, KTabs },
        mocks: {
          $route: {
            params: {
              mesh: 'all',
            },
          },
        },
        stubs: {
          'router-link': RouterLinkStub,
        },
      },
    })

    await screen.findByText(/Mesh: default/)

    await userEvent.click(screen.getByText(/Resources/))

    await screen.findByText(/Rate Limits/)

    expect(container).toMatchSnapshot()
  })
})
