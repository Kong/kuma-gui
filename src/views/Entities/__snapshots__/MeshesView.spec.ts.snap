// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeshesView.vue renders snapshot with Resource tab 1`] = `
<div>
  <div
    class="all-meshes"
  >
    <div
      class="component-frame"
    >
      
      <div
        class="data-overview"
        data-testid="data-overview"
      >
        <div
          class="data-table-controls mb-2"
        >
          
          <a
            class="medium rounded primary k-button add-mesh-button"
            data-v-6bcbf05d=""
            type="button"
          >
            
            <!---->
            
            
            <span
              class="custom-control-icon"
            >
               + 
            </span>
             Create Mesh 
            
            <!---->
          </a>
          
          <button
            class="medium rounded primary k-button refresh-button"
            data-v-6bcbf05d=""
            type="button"
          >
            
            <!---->
            
            
            <span
              class="refresh-icon custom-control-icon"
            >
              <svg
                viewBox="0 0 36 36"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="#fff"
                  fill-rule="nonzero"
                >
                  <path
                    d="M18 5.5a12.465 12.465 0 00-8.118 2.995 1.5 1.5 0 001.847 2.363l.115-.095A9.437 9.437 0 0118 8.5l.272.004a9.487 9.487 0 019.07 7.75l.04.246H25a.5.5 0 00-.416.777l4 6a.5.5 0 00.832 0l4-6 .04-.072A.5.5 0 0033 16.5h-2.601l-.017-.15C29.567 10.2 24.294 5.5 18 5.5zM2.584 18.723l-.04.072A.5.5 0 003 19.5h2.6l.018.15C6.433 25.8 11.706 30.5 18 30.5c3.013 0 5.873-1.076 8.118-2.995a1.5 1.5 0 00-1.847-2.363l-.115.095A9.437 9.437 0 0118 27.5l-.272-.004a9.487 9.487 0 01-9.07-7.75l-.041-.246H11a.5.5 0 00.416-.777l-4-6a.5.5 0 00-.832 0l-4 6z"
                  />
                </g>
              </svg>
            </span>
            <span>
              Refresh
            </span>
            
            <!---->
          </button>
        </div>
        <div
          class="data-overview-content"
        >
          <!-- data -->
          <div
            class="data-overview-table"
          >
            <div
              class=""
              data-testid="data-overview-table"
              data-v-54b66ee0=""
              is-clickable=""
            >
              <section
                class="k-table-wrapper"
                data-v-54b66ee0=""
              >
                <table
                  class="has-hover is-clickable k-table"
                  data-tableid="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                  data-v-54b66ee0=""
                >
                  <thead
                    class=""
                    data-v-54b66ee0=""
                  >
                    <tr
                      class=""
                      data-v-54b66ee0=""
                    >
                      
                      <th
                        class=""
                        data-v-54b66ee0=""
                      >
                        <span
                          class="d-flex align-items-center"
                          data-v-54b66ee0=""
                        >
                          
                          <span
                            class="sr-only"
                            data-v-54b66ee0=""
                          >
                            actions
                          </span>
                          
                          <!---->
                        </span>
                      </th>
                      <th
                        class=""
                        data-v-54b66ee0=""
                      >
                        <span
                          class="d-flex align-items-center"
                          data-v-54b66ee0=""
                        >
                          
                          <span
                            class=""
                            data-v-54b66ee0=""
                          >
                            Name
                          </span>
                          
                          <!---->
                        </span>
                      </th>
                      <th
                        class=""
                        data-v-54b66ee0=""
                      >
                        <span
                          class="d-flex align-items-center"
                          data-v-54b66ee0=""
                        >
                          
                          <span
                            class=""
                            data-v-54b66ee0=""
                          >
                            Type
                          </span>
                          
                          <!---->
                        </span>
                      </th>
                      
                    </tr>
                  </thead>
                  <tbody
                    data-v-54b66ee0=""
                  >
                    
                    <tr
                      class="is-selected"
                      data-v-54b66ee0=""
                      role="link"
                      tabindex="0"
                    >
                      
                      <td
                        class=""
                        data-v-54b66ee0=""
                      >
                        
                        
                        
                      </td>
                      <td
                        class=""
                        data-v-54b66ee0=""
                      >
                        
                        
                        default
                        
                        
                      </td>
                      <td
                        class=""
                        data-v-54b66ee0=""
                      >
                        
                        Mesh
                        
                      </td>
                      
                    </tr>
                    <tr
                      class=""
                      data-v-54b66ee0=""
                      role="link"
                      tabindex="0"
                    >
                      
                      <td
                        class=""
                        data-v-54b66ee0=""
                      >
                        
                        
                        
                      </td>
                      <td
                        class=""
                        data-v-54b66ee0=""
                      >
                        
                        
                        hello-world
                        
                        
                      </td>
                      <td
                        class=""
                        data-v-54b66ee0=""
                      >
                        
                        Mesh
                        
                      </td>
                      
                    </tr>
                    <tr
                      class=""
                      data-v-54b66ee0=""
                      role="link"
                      tabindex="0"
                    >
                      
                      <td
                        class=""
                        data-v-54b66ee0=""
                      >
                        
                        
                        
                      </td>
                      <td
                        class=""
                        data-v-54b66ee0=""
                      >
                        
                        
                        kong-mania-12
                        
                        
                      </td>
                      <td
                        class=""
                        data-v-54b66ee0=""
                      >
                        
                        Mesh
                        
                      </td>
                      
                    </tr>
                    
                  </tbody>
                </table>
                <!---->
              </section>
            </div>
            <div
              class="pagination"
            >
              <!--v-if-->
              <button
                class="medium rounded primary k-button"
                data-v-6bcbf05d=""
                type="button"
              >
                
                <!---->
                
                
                 Next â€º 
                
                <!---->
              </button>
            </div>
          </div>
          <!-- empty state if no items are found -->
          <!--v-if-->
          <!-- additional page content -->
          <!--v-if-->
        </div>
      </div>
      <div
        class="tab-container"
        data-testid="tab-container"
      >
        <header
          class="tab__header"
        >
          
          <div>
            <h3>
               Mesh: default
            </h3>
          </div>
          
        </header>
        <div
          class="tab__content-container"
        >
          <div
            class="k-tabs"
            data-v-6bcd6661=""
          >
            <ul
              aria-label="ktabs"
              data-v-6bcd6661=""
              role="tablist"
            >
              
              <li
                aria-controls="panel-0"
                aria-selected="false"
                class="tab-item"
                data-v-6bcd6661=""
                id="overview-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-6bcd6661=""
                >
                  
                  Overview
                  
                </a>
              </li>
              <li
                aria-controls="panel-1"
                aria-selected="true"
                class="active tab-item"
                data-v-6bcd6661=""
                id="resources-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-6bcd6661=""
                >
                  
                  Resources
                  
                </a>
              </li>
              <li
                aria-controls="panel-2"
                aria-selected="false"
                class="tab-item"
                data-v-6bcd6661=""
                id="yaml-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-6bcd6661=""
                >
                  
                  YAML
                  
                </a>
              </li>
              
            </ul>
            
            <div
              aria-labelledby="overview-tab"
              class="tab-container"
              data-v-6bcd6661=""
              id="panel-0"
              role="tabpanel"
              tabindex="0"
            >
              <!---->
            </div>
            <div
              aria-labelledby="resources-tab"
              class="tab-container"
              data-v-6bcd6661=""
              id="panel-1"
              role="tabpanel"
              tabindex="0"
            >
              
              
              <div>
                <div
                  class="label-list-content"
                >
                  <section
                    aria-describedby="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                    class="noBorder kong-card"
                    data-v-52c90d54=""
                  >
                    <!---->
                    <!---->
                    <div
                      class="k-card-content d-flex"
                      data-v-52c90d54=""
                    >
                      <div
                        class="k-card-body"
                        data-v-52c90d54=""
                        id="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                      >
                        
                        <div
                          class="label-list__col-wrapper multi-col"
                        >
                          
                          
                          <div>
                            <ul>
                              
                              <li>
                                <h4>
                                  Data plane proxies
                                </h4>
                                <p>
                                  10
                                </p>
                              </li>
                              <li>
                                <h4>
                                  Circuit Breakers
                                </h4>
                                <p>
                                  2
                                </p>
                              </li>
                              <li>
                                <h4>
                                  Fault Injections
                                </h4>
                                <p>
                                  2
                                </p>
                              </li>
                              
                            </ul>
                          </div>
                          <div>
                            <ul>
                              
                              <li>
                                <h4>
                                  Health Checks
                                </h4>
                                <p>
                                  4
                                </p>
                              </li>
                              <li>
                                <h4>
                                  Mesh Gateways
                                </h4>
                                <p>
                                  1
                                </p>
                              </li>
                              <li>
                                <h4>
                                  Mesh Gateway Routes
                                </h4>
                                <p>
                                  1
                                </p>
                              </li>
                              
                            </ul>
                          </div>
                          <div>
                            <ul>
                              
                              <li>
                                <h4>
                                  Proxy Templates
                                </h4>
                                <p>
                                  1
                                </p>
                              </li>
                              <li>
                                <h4>
                                  Rate Limits
                                </h4>
                                <p>
                                  0
                                </p>
                              </li>
                              <li>
                                <h4>
                                  Retries
                                </h4>
                                <p>
                                  1
                                </p>
                              </li>
                              
                            </ul>
                          </div>
                          <div>
                            <ul>
                              
                              <li>
                                <h4>
                                  Timeouts
                                </h4>
                                <p>
                                  1
                                </p>
                              </li>
                              <li>
                                <h4>
                                  Traffic Logs
                                </h4>
                                <p>
                                  1
                                </p>
                              </li>
                              <li>
                                <h4>
                                  Traffic Permissions
                                </h4>
                                <p>
                                  3
                                </p>
                              </li>
                              
                            </ul>
                          </div>
                          <div>
                            <ul>
                              
                              <li>
                                <h4>
                                  Traffic Routes
                                </h4>
                                <p>
                                  1
                                </p>
                              </li>
                              <li>
                                <h4>
                                  Traffic Traces
                                </h4>
                                <p>
                                  3
                                </p>
                              </li>
                              <li>
                                <h4>
                                  Virtual Outbounds
                                </h4>
                                <p>
                                  0
                                </p>
                              </li>
                              
                            </ul>
                          </div>
                          
                          
                        </div>
                        
                      </div>
                      <!---->
                    </div>
                  </section>
                </div>
              </div>
              
              
            </div>
            <div
              aria-labelledby="yaml-tab"
              class="tab-container"
              data-v-6bcd6661=""
              id="panel-2"
              role="tabpanel"
              tabindex="0"
            >
              <!---->
            </div>
            
          </div>
          <div>
            <!--v-if-->
          </div>
        </div>
      </div>
      
    </div>
  </div>
</div>
`;
